pkg/store/multitsdb.go

```diff
@@ -0,0 +1,275 @@
+		}
+		infos = append(infos, info)
+	}
+
+	resp.MinTime = infos[0].MinTime
+	resp.MaxTime = infos[0].MaxTime
+
+	for i := 1; i < len(infos); i++ {
+		if resp.MinTime > infos[i].MinTime {
+			resp.MinTime = infos[i].MinTime
+		}
+		if resp.MaxTime < infos[i].MaxTime {
+			resp.MaxTime = infos[i].MaxTime
+		}
+	}
+
+	// We can rely on every underlying TSDB to only have one labelset, so this
+	// will always allocate the correct length immediately.
+	resp.LabelSets = make([]storepb.LabelSet, 0, len(infos))
+	for _, info := range infos {
+		resp.LabelSets = append(resp.LabelSets, info.LabelSets...)
+	}
+
+	return resp, nil
+}
+
+type seriesSetServer struct {
+	grpc.ServerStream
+
+	ctx context.Context
+
+	warnCh warnSender
+	recv   chan *storepb.Series
+	cur    *storepb.Series
+
+	err error
+}
+
+func newSeriesSetServer(
+	ctx context.Context,
+	warnCh warnSender,
+) *seriesSetServer {
+	return &seriesSetServer{
+		ctx:    ctx,
+		warnCh: warnCh,
+		recv:   make(chan *storepb.Series),
+	}
+}
+
+func (s *seriesSetServer) Context() context.Context {
+	return s.ctx
+}
+
+func (s *seriesSetServer) Series(store *TSDBStore, r *storepb.SeriesRequest) {
+	err := store.Series(r, s)
+	if err != nil {
+		if r.PartialResponseDisabled {
+			s.err = err
+		} else {
+			s.warnCh.send(storepb.NewWarnSeriesResponse(err))
+		}
+	}
+
+	close(s.recv)
+}
+
+func (s *seriesSetServer) Send(r *storepb.SeriesResponse) error {
+	series := r.GetSeries()
+	chunks := make([]storepb.AggrChunk, len(series.Chunks))
+	copy(chunks, series.Chunks)
+	s.recv <- &storepb.Series{
+		Labels: series.Labels,
+		Chunks: chunks,
+	}
+	return nil
+}
+
+func (s *seriesSetServer) Next() (ok bool) {
+	s.cur, ok = <-s.recv
+	return ok
+}
+
+func (s *seriesSetServer) At() ([]storepb.Label, []storepb.AggrChunk) {
+	if s.cur == nil {
+		return nil, nil
+	}
+	return s.cur.Labels, s.cur.Chunks
+}
+
+func (s *seriesSetServer) Err() error {
+	return s.err
+}
+
+// Series returns all series for a requested time range and label matcher. The
+// returned data may exceed the requested time bounds. The data returned may
+// have been read and merged from multiple underlying TSDBStore instances.
+func (s *MultiTSDBStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+	stores := s.tsdbStores()
+	if len(stores) == 0 {
+		return nil
+	}
+
+	var (
+		g, gctx = errgroup.WithContext(srv.Context())
+
+		// Allow to buffer max 10 series response.
+		// Each might be quite large (multi chunk long series given by sidecar).
+		respSender, respRecv, closeFn = newRespCh(gctx, 10)
+	)
+
+	g.Go(func() error {
+		var (
+			seriesSet []storepb.SeriesSet
+			wg        = &sync.WaitGroup{}
+		)
+
+		defer func() {
+			wg.Wait()
+			closeFn()
+		}()
+
+		for tenant, store := range stores {
+			store := store
+			seriesCtx, closeSeries := context.WithCancel(gctx)
+			seriesCtx = grpc_opentracing.ClientAddContextTags(seriesCtx, opentracing.Tags{
+				"tenant": tenant,
+			})
+			defer closeSeries()
```

Comment:
```suggestion
			defer cancelSeries()
```

As we usually cancel context not close (: